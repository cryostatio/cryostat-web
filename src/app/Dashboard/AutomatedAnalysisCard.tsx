/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import { BreadcrumbPage, BreadcrumbTrail } from '@app/BreadcrumbPage/BreadcrumbPage';
import { ServiceContext } from '@app/Shared/Services/Services';
import { TargetSelect } from '@app/TargetSelect/TargetSelect';
import { Card, CardActions, CardBody, CardHeader, CardTitle, Checkbox, Dropdown, Grid, GridItem, KebabToggle, Label, LabelGroup, LabelProps, Stack, StackItem, Tooltip } from '@patternfly/react-core';
import { useSubscriptions } from '@app/utils/useSubscriptions';
import { NO_TARGET } from '@app/Shared/Services/Target.service';
import { InfoCircleIcon, RulerVerticalIcon } from '@patternfly/react-icons';
import { number } from 'prop-types';

interface AutomatedAnalysisCardProps {
  pageTitle: string;
  compactSelect?: boolean;
  hideEmptyState?: boolean;
  breadcrumbs?: BreadcrumbTrail[];
}

export const AutomatedAnalysisCard: React.FunctionComponent<AutomatedAnalysisCardProps> = (props) => {
  const context = React.useContext(ServiceContext);
  const addSubscription = useSubscriptions();

  const [categorizedEvaluation, setCategorizedEvaluation] = React.useState<[string , RuleEvaluation[]][]>([] as [string, RuleEvaluation[]][]);
  const [criticalCategorizedEvaluation, setCriticalCategorizedEvaluation] = React.useState<[string , RuleEvaluation[]][]>([] as [string, RuleEvaluation[]][]);
  const [ruleLabels, setRuleLabels] = React.useState<RuleEvaluation[]>([]);
  const [isOpen, setIsOpen] = React.useState<boolean>(false);
  const [isChecked, setIsChecked] = React.useState<boolean>(false);

  const categorizeEvaluation = React.useCallback((arr: [string, RuleEvaluation][]) => {
    const map = new Map<string, RuleEvaluation[]>();
    arr.forEach(([name, evaluation]) => {
      const obj = map.get(evaluation.topic);
      if (obj === undefined) {
          map.set(evaluation.topic, [evaluation]);
      }
      else {
          obj.push(evaluation);
      }
    });
    setCategorizedEvaluation(Array.from(map) as [string, RuleEvaluation[]][]);
  }, [setCategorizedEvaluation]);

  const takeSnapshot = React.useCallback(() => {
    addSubscription(
        context.api.createSnapshotv2().subscribe((snapshot) => {
            addSubscription(
              context.reports.reportJson(snapshot)
                .subscribe((report) => {
                  const reportJson = JSON.parse(report);
                  const arr: [string, RuleEvaluation][] = Object.entries(reportJson);
                  categorizeEvaluation(arr);
                })
            )
        })
    );
  }, [addSubscription, context.api, categorizeEvaluation]);

  const showCriticalScores = React.useCallback(() => {
    console.log(categorizedEvaluation);
    const criticalScores = categorizedEvaluation.filter(d => d[1].some(e => e.score >= 50));
    setCriticalCategorizedEvaluation(criticalScores);
  }, [categorizedEvaluation, setCriticalCategorizedEvaluation]);

  React.useEffect(() => {
    takeSnapshot();
  }, [takeSnapshot]);

  React.useEffect(() => {
    if (isChecked) {
      showCriticalScores();
    }
    else {
      setCriticalCategorizedEvaluation(categorizedEvaluation);
    }
  }, [isChecked, showCriticalScores, setCriticalCategorizedEvaluation]);


  const onSelect = () => {
    setIsOpen(!isOpen);
  };
  const onClick = (checked: boolean) => {
    setIsChecked(checked);
  };

  const filteredCategorizedLabels = React.useMemo(() => {
    return criticalCategorizedEvaluation.map(([topic, evaluations]) => {
        return (
            <LabelGroup 
              categoryName={topic}
              isVertical
              numLabels={4}
              isCompact
              key={topic}
            >
              {
                evaluations
                  .filter((evaluation) => evaluation.score >= 0)
                  .map((evaluation) => {
                  return (
                    <Tooltip content={`${evaluation.name} Score:  ${evaluation.score}`} key={evaluation.name}>
                      <Label icon={<InfoCircleIcon />} color={evaluation.score < 10 ? 'green' : (evaluation.score < 50) ? 'orange' : 'red'} isCompact >
                        {evaluation.name}
                      </Label>
                    </Tooltip>
                  )
                })
              }
            </LabelGroup>
        )
      });
  }, [criticalCategorizedEvaluation]);

  interface IRule {
    name: string;
    evaluation: RuleEvaluation;
  }

  interface RuleEvaluation {
    name: string;
    description: string;
    score: number;
    topic: string;
  }


  return (
        <Card isRounded isCompact>
          <CardHeader>
            <CardTitle component='h4'>Automated Analysis</CardTitle>
            <CardActions>
              <Dropdown
                onSelect={onSelect}
                toggle={<KebabToggle onToggle={setIsOpen} />}
                isOpen={isOpen}
                isPlain
                dropdownItems={[]}
                position={'right'}
              />
              <Checkbox
                isChecked={isChecked}
                onChange={onClick}
                aria-label="card checkbox example"
                id="check-2"
                name="check2"
              />
            </CardActions>
        </CardHeader>
            <CardBody isFilled={true}>
              {filteredCategorizedLabels}
            </CardBody>
        </Card>
  );
};
