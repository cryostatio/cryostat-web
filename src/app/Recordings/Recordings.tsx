/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ServiceContext } from '@app/Shared/Services/Services';
import { TargetView } from '@app/TargetView/TargetView';
import { useSubscriptions } from '@app/utils/useSubscriptions';
import { getActiveTab, switchTab } from '@app/utils/utils';
import { Card, CardBody, CardTitle, Tab, Tabs, TabTitleText } from '@patternfly/react-core';
import * as React from 'react';
import { useHistory, useLocation } from 'react-router';
import { ActiveRecordingsTable } from './ActiveRecordingsTable';
import { ArchivedRecordingsTable } from './ArchivedRecordingsTable';

enum RecordingTab {
  ACTIVE_RECORDING = 'active-recording',
  ARCHIVED_RECORDING = 'archived-recording',
}

export interface RecordingsProps {}

export const Recordings: React.FC<RecordingsProps> = ({ ...props }) => {
  const { search, pathname } = useLocation();
  const history = useHistory();
  const context = React.useContext(ServiceContext);
  const addSubscription = useSubscriptions();

  const activeTab = React.useMemo(() => {
    return getActiveTab(search, 'tab', Object.values(RecordingTab), RecordingTab.ACTIVE_RECORDING);
  }, [search]);

  const [archiveEnabled, setArchiveEnabled] = React.useState(false);

  React.useEffect(() => {
    addSubscription(context.api.isArchiveEnabled().subscribe(setArchiveEnabled));
  }, [context.api, addSubscription, setArchiveEnabled]);

  const onTabSelect = React.useCallback(
    (_: React.MouseEvent, key: string | number) =>
      switchTab(history, pathname, search, { tabKey: 'tab', tabValue: `${key}` }),
    [history, pathname, search],
  );

  const targetAsObs = React.useMemo(() => context.target.target(), [context.target]);

  const cardBody = React.useMemo(() => {
    return archiveEnabled ? (
      <Tabs id="recordings" activeKey={activeTab} onSelect={onTabSelect} unmountOnExit>
        <Tab
          id="active-recordings"
          eventKey={RecordingTab.ACTIVE_RECORDING}
          title={<TabTitleText>Active Recordings</TabTitleText>}
          data-quickstart-id="active-recordings-tab"
        >
          <ActiveRecordingsTable archiveEnabled={true} />
        </Tab>
        <Tab
          id="archived-recordings"
          eventKey={RecordingTab.ARCHIVED_RECORDING}
          title={<TabTitleText>Archived Recordings</TabTitleText>}
          data-quickstart-id="archived-recordings-tab"
        >
          <ArchivedRecordingsTable target={targetAsObs} isUploadsTable={false} isNestedTable={false} />
        </Tab>
      </Tabs>
    ) : (
      <>
        <CardTitle>Active Recordings</CardTitle>
        <ActiveRecordingsTable archiveEnabled={false} />
      </>
    );
  }, [archiveEnabled, activeTab, onTabSelect, targetAsObs]);

  return (
    <TargetView {...props} pageTitle="Recordings">
      <Card>
        <CardBody>{cardBody}</CardBody>
      </Card>
    </TargetView>
  );
};

export default Recordings;
