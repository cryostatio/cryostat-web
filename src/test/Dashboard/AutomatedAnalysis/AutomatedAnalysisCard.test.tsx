/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import { AutomatedAnalysisCard } from '@app/Dashboard/AutomatedAnalysis/AutomatedAnalysisCard';
import { emptyActiveRecordingFilters, emptyArchivedRecordingFilters } from '@app/Recordings/RecordingFilters';
import { TargetRecordingFilters } from '@app/Shared/Redux/RecordingFilterReducer';
import { RootState } from '@app/Shared/Redux/ReduxStore';
import {
  ActiveRecording,
  ArchivedRecording,
  AutomatedAnalysisRecordingConfig,
  defaultAutomatedAnalysisRecordingConfig,
} from '@app/Shared/Services/Api.service';
import {
  CachedReportValue,
  FAILED_REPORT_MESSAGE,
  NO_RECORDINGS_MESSAGE,
  RuleEvaluation,
} from '@app/Shared/Services/Report.service';
import { defaultServices } from '@app/Shared/Services/Services';
import { automatedAnalysisConfigToRecordingAttributes } from '@app/Shared/Services/Settings.service';
import '@testing-library/jest-dom';
import { cleanup, screen } from '@testing-library/react';
import * as React from 'react';
import { of } from 'rxjs';
import { renderWithServiceContextAndReduxStore } from '../../Common';

const mockConnectUrl = 'service:jmx:rmi://someUrl';
const mockTarget = { connectUrl: mockConnectUrl, alias: 'fooTarget' };

const mockEmptyCachedReport: CachedReportValue = {
  report: [],
  timestamp: 0,
};

const mockEvaluations: RuleEvaluation[] = [
  {
    name: 'rule1',
    description: 'rule1 description',
    score: 100,
    topic: 'rule1 topic',
  },
];

const mockRecording = {
  name: 'someRecording',
};

const mockArchivedRecording: ArchivedRecording = {
  name: 'someArchivedRecording',
  downloadUrl: '',
  reportUrl: '',
  metadata: { labels: {} },
  size: 0,
  archivedTime: 1663027200000, // 2022-09-13T00:00:00.000Z in milliseconds
};

const mockRecordingConfig: AutomatedAnalysisRecordingConfig = {
  templates: '',
  maxSize: 0,
  maxAge: 0,
};

const mockTargetNode = {
  recordings: {
    active: {
      data: [mockRecording],
    },
  },
};

const mockActiveRecordingsResponse = {
  data: {
    targetNodes: [mockTargetNode],
  },
};

const mockEmptyActiveRecordingsResponse = {
  data: {
    targetNodes: [
      {
        recordings: {
          active: {
            data: [],
          },
        },
      },
    ],
  },
};

const mockArchivedRecordingsResponse = {
  data: {
    archivedRecordings: {
      data: [mockArchivedRecording],
    },
  },
};

const mockEmptyArchivedRecordingsResponse = {
  data: {
    archivedRecordings: {
      data: [],
    },
  },
};

jest.mock('@app/Dashboard/AutomatedAnalysis/AutomatedAnalysisFilters', () => {
  return {
    ...jest.requireActual('@app/Dashboard/AutomatedAnalysis/AutomatedAnalysisFilters'),
    RecordingFilters: jest.fn(() => {
      return <div>AutomatedAnalysisFilters</div>;
    }),
  };
});

jest.mock('@app/Dashboard/AutomatedAnalysis/Filters/AutomatedAnalysisScoreFilter', () => {
  return {
    ...jest.requireActual('@app/Dashboard/AutomatedAnalysis/Filters/AutomatedAnalysisScoreFilter'),
    RecordingFilters: jest.fn(() => {
      return <div>Score Filter</div>;
    }),
  };
});

jest.mock('@app/Dashboard/AutomatedAnalysis/AutomatedAnalysisConfigDrawer', () => {
  return {
    ...jest.requireActual('@app/Dashboard/AutomatedAnalysis/AutomatedAnalysisConfigDrawer'),
    RecordingFilters: jest.fn(() => {
      return <div>AutomatedAnalysisConfigDrawer</div>;
    }),
  };
});

jest.spyOn(defaultServices.target, 'target').mockReturnValue(of(mockTarget));
jest.spyOn(defaultServices.target, 'authFailure').mockReturnValue(of());

describe('<AutomatedAnalysisCard />', () => {
  let preloadedState: RootState;

  beforeEach(() => {
    preloadedState = {
      recordingFilters: {
        list: [
          {
            target: mockTarget.connectUrl,
            active: {
              selectedCategory: 'Labels',
              filters: emptyActiveRecordingFilters,
            },
            archived: {
              selectedCategory: 'Name',
              filters: emptyArchivedRecordingFilters,
            },
          } as TargetRecordingFilters,
        ],
      },
      automatedAnalysisFilters: {
        state: {
          targetFilters: [],
          globalFilters: {
            filters: {
              Score: 100,
            },
          },
        },
      },
    };
  });

  afterEach(() => {
    cleanup();
    jest.useRealTimers();
  });

  // it('renders correctly', async () => {
  //   const myObservable = timer(0, 200);

  //   jest.spyOn(defaultServices.api, 'graphql').mockReturnValueOnce(of());
  //   jest.spyOn(defaultServices.api, 'graphql').mockReturnValueOnce(of());

  //   jest.spyOn(defaultServices.reports, 'reportJson').mockReturnValueOnce(of());
  //   jest.spyOn(defaultServices.reports, 'getCachedAnalysisReport').mockReturnValueOnce(mockEmptyCachedReport);

  //   renderWithServiceContextAndReduxStore(<AutomatedAnalysisCard pageTitle={'Automated Analysis'} />, {
  //     preloadState: preloadedState,
  //   });

  //   expect(screen.getByText('Automated Analysis')).toBeInTheDocument(); // Card title
  //   expect(screen.getByLabelText('Details')).toBeInTheDocument(); // Expandable content button
  //   expect(screen.getByText('Name')).toBeInTheDocument(); // Default state filter
  //   expect(screen.getByLabelText('Refresh automated analysis')).toBeInTheDocument(); // Refresh button
  //   expect(screen.getByLabelText('Delete automated analysis')).toBeInTheDocument(); // Delete button
  //   expect(screen.getByText('Show N/A scores')).toBeInTheDocument();
  //   expect(screen.getByLabelText('automated-analysis-toolbar')).toBeInTheDocument(); // Toolbar

  //   expect(screen.getByText("Loading")).toBeInTheDocument(); // Loading view

  //   // jest.advanceTimersByTime(0);

  //   myObservable.subscribe(() => {
  //     console.log("SFD")

  //     expect(screen.getByText("Loading")).toBeInTheDocument(); // Loading view

  //   });

  // });

  it('renders report generation error view correctly', async () => {
    jest.spyOn(defaultServices.api, 'graphql').mockReturnValueOnce(of(mockActiveRecordingsResponse));
    jest.spyOn(defaultServices.reports, 'reportJson').mockReturnValueOnce(of());
    renderWithServiceContextAndReduxStore(<AutomatedAnalysisCard pageTitle={'Automated Analysis'} />, {
      preloadState: preloadedState,
    });

    expect(screen.getByText('Automated Analysis Error')).toBeInTheDocument(); // Error view
    expect(screen.getByText(FAILED_REPORT_MESSAGE)).toBeInTheDocument(); // Error message
    expect(screen.getByText('Cryostat was unable to generate an automated analysis report.')).toBeInTheDocument(); // Error details
    expect(screen.getByRole('button', { name: /retry loading report/i })).toBeInTheDocument(); // Retry button
    expect(screen.queryByLabelText('automated-analysis-toolbar')).not.toBeInTheDocument(); // Toolbar
  });

  it('renders empty recordings error view and creates recording when clicked', async () => {
    jest.spyOn(defaultServices.api, 'graphql').mockReturnValueOnce(of(mockEmptyActiveRecordingsResponse));
    jest.spyOn(defaultServices.reports, 'getCachedAnalysisReport').mockReturnValueOnce(mockEmptyCachedReport);
    jest.spyOn(defaultServices.api, 'graphql').mockReturnValueOnce(of(mockEmptyArchivedRecordingsResponse));

    jest.spyOn(defaultServices.api, 'createRecording').mockReturnValueOnce(of());
    const { user } = renderWithServiceContextAndReduxStore(<AutomatedAnalysisCard pageTitle={'Automated Analysis'} />, {
      preloadState: preloadedState,
    });

    const requestSpy = jest.spyOn(defaultServices.api, 'createRecording');

    expect(screen.getByText('Automated Analysis Error')).toBeInTheDocument(); // Error view
    expect(screen.getByText(NO_RECORDINGS_MESSAGE)).toBeInTheDocument(); // Error message
    expect(screen.getByText('Cryostat was unable to generate an automated analysis report.')).toBeInTheDocument(); // Error details
    expect(screen.queryByLabelText('automated-analysis-toolbar')).not.toBeInTheDocument(); // Toolbar

    await user.click(screen.getByRole('button', { name: /recording config dropdown/i }));

    const defaultMenuItem = screen.getByRole('menuitem', {
      name: /default/i,
    });
    const customMenuItem = screen.getByRole('menuitem', {
      name: /custom/i,
    });

    expect(defaultMenuItem).toBeInTheDocument(); // Default recording config
    expect(customMenuItem).toBeInTheDocument(); // Custom recording config

    await user.click(defaultMenuItem);

    expect(requestSpy).toHaveBeenCalledTimes(1);
    expect(requestSpy).toBeCalledWith(
      automatedAnalysisConfigToRecordingAttributes(defaultAutomatedAnalysisRecordingConfig)
    );
  });

  it('renders archived recordings', async () => {
    const mockCurrentDate = new Date('14 Sep 2022 00:00:00 UTC');
    jest.useFakeTimers('modern').setSystemTime(mockCurrentDate);
    jest.spyOn(defaultServices.api, 'graphql').mockReturnValueOnce(of(mockEmptyActiveRecordingsResponse));
    jest.spyOn(defaultServices.reports, 'getCachedAnalysisReport').mockReturnValueOnce(mockEmptyCachedReport);
    jest.spyOn(defaultServices.api, 'graphql').mockReturnValueOnce(of(mockArchivedRecordingsResponse));

    jest.spyOn(defaultServices.reports, 'reportJson').mockReturnValueOnce(of(mockEvaluations));
    jest.spyOn(defaultServices.api, 'createRecording').mockReturnValueOnce(of());
    const { user } = renderWithServiceContextAndReduxStore(<AutomatedAnalysisCard pageTitle={'Automated Analysis'} />, {
      preloadState: preloadedState,
    });
    const requestSpy = jest.spyOn(defaultServices.api, 'createRecording');
    screen.logTestingPlaygroundURL();

    expect(screen.getByText('Automated Analysis')).toBeInTheDocument(); // Card title
    expect(screen.getByLabelText('Details')).toBeInTheDocument(); // Expandable content button
    expect(screen.getByText('Name')).toBeInTheDocument(); // Default state filter
    expect(screen.getByLabelText('Refresh automated analysis')).toBeInTheDocument(); // Refresh button
    expect(screen.getByLabelText('Delete automated analysis')).toBeInTheDocument(); // Delete button
    expect(screen.getByText('Show N/A scores')).toBeInTheDocument();
    expect(screen.getByLabelText('automated-analysis-toolbar')).toBeInTheDocument(); // Toolbar

    const createButton = screen.getByRole('button', {
      name: /create default recording/i,
    });
    expect(createButton).toBeInTheDocument(); // Loading view
    expect(screen.getByText('Showing archived report from from 1 day ago.')).toBeInTheDocument(); // Loading view

    jest.useRealTimers();

    await user.click(createButton);

    expect(requestSpy).toHaveBeenCalledTimes(1);
    expect(requestSpy).toBeCalledWith(
      automatedAnalysisConfigToRecordingAttributes(defaultAutomatedAnalysisRecordingConfig)
    );
  });

  //   it('should show error view if failing to retrieve recordings', async () => {
  //     const subj = new Subject<void>();
  //     const mockTargetSvc = {
  //       target: () => of(mockTarget),
  //       authFailure: () => subj.asObservable(),
  //     } as TargetService;
  //     const services: Services = {
  //       ...defaultServices,
  //       target: mockTargetSvc,
  //     };

  //     renderWithServiceContextAndReduxStore(<ActiveRecordingsTable archiveEnabled={true} />, {
  //       preloadState: preloadedState,
  //       services,
  //     });

  //     await act(async () => subj.next());

  //     const failTitle = screen.getByText('Error retrieving recordings');
  //     expect(failTitle).toBeInTheDocument();
  //     expect(failTitle).toBeVisible();

  //     const authFailText = screen.getByText('Auth failure');
  //     expect(authFailText).toBeInTheDocument();
  //     expect(authFailText).toBeVisible();

  //     const retryButton = screen.getByText('Retry');
  //     expect(retryButton).toBeInTheDocument();
  //     expect(retryButton).toBeVisible();

  //     const toolbar = screen.queryByLabelText('active-recording-toolbar');
  //     expect(toolbar).not.toBeInTheDocument();
  //   });
});
