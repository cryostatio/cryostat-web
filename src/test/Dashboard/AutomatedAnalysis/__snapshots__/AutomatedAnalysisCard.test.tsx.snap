// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutomatedAnalysisCard /> renders correctly 1`] = `
<div
  className="draggable-ref-wrapper"
  onDragStart={[Function]}
  onTransitionEnd={[Function]}
  style={Object {}}
>
  <div
    className="dashboard-card-resizable-wrapper"
  >
    <div
      className="pf-v5-c-card pf-m-compact pf-m-expanded pf-m-rounded dashboard-card"
      data-ouia-component-id="OUIA-Generated-Card-1"
      data-ouia-component-type="PF5/Card"
      data-ouia-safe={true}
      id="automated-analysis-card"
    >
      <div
        className="draggable-ref__grip"
        data-quickstart-id="card-draggable-grip"
        draggable={true}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="pf-v5-c-card__header"
        >
          <div
            className="pf-v5-c-card__header-toggle"
          >
            <button
              aria-disabled={false}
              aria-expanded={true}
              aria-label="Details"
              aria-labelledby="automated-analysis-card-title toggle-details"
              className="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe={true}
              disabled={false}
              id="automated-analysis-toggle-details"
              onClick={[Function]}
              role={null}
              type="button"
            >
              <span
                className="pf-v5-c-card__header-toggle-icon"
              >
                <svg
                  aria-hidden="true"
                  aria-labelledby={null}
                  className="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            className="pf-v5-c-card__actions"
          >
            <div
              style={
                Object {
                  "display": "contents",
                }
              }
            >
              <button
                aria-disabled={false}
                aria-label="refresh"
                className="pf-v5-c-button pf-m-plain"
                data-ouia-component-id="OUIA-Generated-Button-plain-2"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe={true}
                disabled={false}
                onClick={[Function]}
                role={null}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  aria-labelledby={null}
                  className="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 1024 1024"
                  width="1em"
                >
                  <path
                    d="M1016,640.1 L724,640.1 L695.9,608.7 C701.8,597.3 706.7,585.2 710.5,572.7 L793.5,567.2 C799.9,566.3 804.9,559.9 804.9,553.5 L804.9,468.2 C804.8,461.7 800.2,456.2 793.8,454.4 L709.3,449.7 C705.5,437.7 700.5,426.1 694.7,415.1 L750.4,351.5 C754.3,346.3 753.3,338.3 748.8,333.8 L688.5,273.5 C684,269 676.8,268.3 671,271.6 L607.3,328.6 C596.4,322.9 585,318.2 573.2,314.5 L567.6,229.6 C566.7,223.2 560.3,218.2 553.9,218.2 L468.6,218.2 C462.2,218.2 456.7,222.8 454.9,229.2 L450.2,314.2 C438,317.9 426.3,322.8 415.2,328.6 L351.1,273.6 C345.9,269.7 337.9,270.7 333.4,275.2 L304.3,304.3 L240.6,240.6 C310.1,171.1 406.1,128 512,128 C718.3,128 887.2,291.6 895.6,495.9 C896,504.5 903.1,511.2 911.6,511.2 L1007.7,511.2 C1016.7,511.2 1024,503.7 1023.7,494.7 C1014.6,220.4 788.6,0 512.1,0 C370.9,0 242.9,57.5 150.2,150.1 L13.7,13.6 C8.6,8.6 0,12.2 0,19.3 L0,376.1 C0,380.5 3.6,384.1 8,384.1 L299.9,384.1 L328.1,414.6 C322,426 316.9,438 312.9,450.5 L230.1,456.9 C223.7,457.8 218.7,464.2 218.7,470.6 L218.7,555.9 C218.7,562.3 223.3,567.8 229.7,569.6 L312,572.9 C315.9,585.9 321,598.3 327.2,610.1 L274,672.1 C270.1,677.3 271.1,685.3 275.6,689.8 L335.9,750.1 C340.4,754.6 347.6,755.3 353.4,752 L413.3,696.7 C425.2,703.1 437.9,708.3 451,712.3 L457.2,793.1 C458.1,799.5 464.5,804.5 470.9,804.5 L556.2,804.5 C562.6,804.5 568.1,799.9 569.9,793.5 L573.1,711.7 C585.9,707.7 598.2,702.5 609.9,696.2 L671.4,750.1 C676.6,754 684.6,753 689.1,748.5 L718.9,718.7 L783.6,783.4 C714,852.9 618,896 512.1,896 C305.4,896 136.4,732.1 128.4,527.3 C128,518.7 121,512 112.4,512 L16.3,512 C7.3,512 0,519.5 0.3,528.5 C9,803.2 235.3,1024 512.1,1024 C653.2,1024 781.2,966.6 873.9,873.9 L1010.3,1010.4 C1015.4,1015.4 1024,1011.8 1024,1004.7 L1024,648 C1024,643.6 1020.4,640 1016,640.1 M511.1,640.9 C440.4,640.9 383.1,583.6 383.1,512.9 C383.1,442.2 440.4,384.9 511.1,384.9 C581.8,384.9 639.1,442.2 639.1,512.9 C639.1,583.6 581.8,640.9 511.1,640.9"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            className="pf-v5-c-card__header-main"
          >
            <div
              className="pf-v5-l-flex"
            >
              <div
                className=""
              >
                <div
                  className="pf-v5-c-card__title"
                >
                  <h4
                    className="pf-v5-c-card__title-text"
                    id="automated-analysis-card-title"
                  >
                    Automated analysis
                  </h4>
                </div>
              </div>
              <div
                className=""
              >
                <div
                  className="pf-v5-c-label-group"
                >
                  <div
                    className="pf-v5-c-label-group__main"
                  >
                    
                    <ul
                      aria-label="Label group category"
                      className="pf-v5-c-label-group__list"
                      role="list"
                    >
                      <li
                        className="pf-v5-c-label-group__list-item"
                      >
                        <span
                          className="pf-v5-c-label pf-m-orange"
                          draggable={true}
                          onDragStart={[Function]}
                          onFocus={[Function]}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                        >
                          <button
                            className="pf-v5-c-label__content"
                            onClick={[Function]}
                            type="button"
                          >
                            <span
                              className="pf-v5-c-label__icon"
                            >
                              <svg
                                aria-hidden={true}
                                aria-labelledby={null}
                                className="pf-v5-svg"
                                fill="currentColor"
                                height="1em"
                                role="img"
                                viewBox="0 0 576 512"
                                width="1em"
                              >
                                <path
                                  d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                                />
                              </svg>
                            </span>
                            <span
                              className="pf-v5-c-label__text"
                            >
                              1 warning result
                            </span>
                          </button>
                        </span>
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="pf-v5-c-card__expandable-content"
      >
        <div
          className="pf-v5-c-card__body"
        >
          <div>
            TargetAnalysis
          </div>
        </div>
      </div>
    </div>
    <div
      className="resizable-ref"
      onMouseDown={[Function]}
    />
  </div>
</div>
`;
